scanner/decisions/0010-separate-versioning-for-vuln-bundles.md

```diff
@@ -0,0 +1,84 @@
+
+This approach needs to be revised for the following reasons:
+
+1.  The vulnerability schema rarely changes, but maintaining multiple bundles is expensive.  It requires multiple GitHub workflow runs, which increases the chances of failure and pressure on the security sources' endpoints.
+2.  It can be challenging to backport fixes and updates to previous releases.  This requires patch releases, which trigger the release workflow.  Improving the release process can be complex, as it may require patch release cuts, changes to different branches, and integration between logic that is managed in the master branch, executing other releases binaries.
+3.  The latest offline bundles does not support post `4.5` releases.  Bundle embedding for each release increases the cost to store and consume offline bundles.  The `roxctl` interface and documentation today needs to be clearer on these limitations.
+
+## Decision
+
+We will implement a versioning stream for vulnerability bundles, bumped when we determine that ClairCore's schema has changed.
+
+Initially, two version streams will exist (table 1 below):
+
+| Stream | Git Ref      | Description                                                        |
+|--------|--------------|--------------------------------------------------------------------|
+| dev    | heads/master | This stream tracks the latest vulnerability schema                 |
+| v1     | heads/4.4.5  | This stream tracks the schema used by the current StackRox release |
+
+On the subsequent releases (for example,. `4.4.6`), one of the two will be carried out:
+
+1.  The schema hasn't changed, so v1 now tracks the new release (`heads/4.4.6`).
+2.  The schema has changed, so v1 stays at the current release (`heads/4.4.5`), and a new version, v2, is created to track the next release (`heads/4.4.6`).
+
+### Building versioned bundles
+
+The GitHub workflow will have a map with "bundle version" as keys and "git reference" as values, similar to Table 1 above.  This can be stored statically in the matrix definition.
+
+For each version in the mapping, a bundle will be built based on the specified git ref and stored at `https://definitions.stackrox.io/v4/<bundle-version>/vulnerabilities.zip`.
+
+In every release cut, the key specified in `scanner/VULNERABILITY_VERSION` will be updated in the mapping to the current release branch reference (e.g., `heads/4.4.6`).  Notice that if the version is the same, it will be updated.  If it's new, it will be added, and the previous version will stay.
```

Comment:
> In every release cut, the key specified in `scanner/VULNERABILITY_VERSION` will be updated in the mapping to the current release branch reference (e.g., `heads/4.4.6`).

Based on the description I would assume `VULNERABILITY_VERSION` would contain the bundle version (i.e. `v1`) instead of the ref (i.e. `heads/4.4.6`), like this:

branch | `scanner/VULNERABILITY_VERSION` content
--- | ---
`master` | `dev`
`release-4.4` | `v1`
`release-4.5` | `v1`

The example however mentions "(e.g., `heads/4.4.6`)", could you clarify / provide an example of what the contents of this file will be for various branches, tags, etc.?  If the file contains the `ref`, I'm a bit confused how this improves current state, such that:

ACS version `4.4.6` was released with `VULNERABILITY_VERSION` = `heads/4.4.6`
ACS version `4.4.7` was released with `VULNERABILITY_VERSION` = `heads/4.4.7`
and so on.

Reply:
Let me clarify:

> In every release cut, the key specified in `scanner/VULNERABILITY_VERSION` will be updated in the mapping to the current release branch reference (e.g., `heads/4.4.6`).  Notice that if the version is the same, it will be updated.  If it's new, it will be added, and the previous version will stay.


In a release cut (let's say 4.6) we will look into `scanner/VULNERABILITY_VERSION`, let's say its contents are:

```
$ cat canner/VULNERABILITY_VERSION
v1
```

So, now I will update the mapping to the current release branch reference. Let's say the mapping was:

bundle version | git reference
--- | ---
`dev` | `master`
`v1` | `heads/release-4.5`

So I will update `v1` to the new release (in this example `4.6`):

bundle version | git reference
--- | ---
`dev` | `master`
`v1` | `heads/release-4.6`

If in 4.6 we updated the bundle version, the content of the file would be:

```
$ cat canner/VULNERABILITY_VERSION
v2
```

Then the final table would be:

bundle version | git reference
--- | ---
`dev` | `master`
`v1` | `heads/release-4.5`
`v2` | `heads/release-4.6`

Reply:
That makes sense now, can you add some details regarding the mapping? Where will that be stored and what format?

Reply:
That is intentionally open, it could be hard-coded in the YAML workflow, or a separate file. It depends on how and where the mapping between releases and bundle versions are needed.

Reply:
I added this example to the ADR.